pkg/webhook/podmarker/validating/podmarker_create_update_handler.go

```diff
@@ -0,0 +1,188 @@
+		replicas, err := intstr.GetValueFromIntOrPercent(strategy.Replicas, 100, true)
+		if err != nil {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("replicas"), err))
+		} else if replicas < 0 || replicas > 100 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("replicas"), *strategy.Replicas, "replicas must be a non-negative integer, or a percentage satisfying 0% <= replicas <= 100%"))
+		}
+	}
+	return
+}
+
+func validatePodMarkerMatchRequirements(requirement *appsv1alpha1.PodMarkerRequirements, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. validate podSelector
+	podSelector, err := v1.LabelSelectorAsSelector(requirement.PodSelector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("podSelector"), requirement.PodSelector, fmt.Sprintf("invalid podSelector, err: %v", err)))
+	}
+	// 2. validate nodeSelector
+	nodeSelector, err := v1.LabelSelectorAsSelector(requirement.NodeSelector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("nodeSelector"), requirement.NodeSelector, fmt.Sprintf("invalid NodeSelector, err: %v", err)))
+	}
+	// MatchRequirements cannot be empty
+	podRequirements, _ := podSelector.Requirements()
+	nodeRequirements, _ := nodeSelector.Requirements()
+	if len(podRequirements)+len(nodeRequirements) == 0 {
+		return append(allErrs, field.Invalid(fldPath, *requirement, fmt.Sprintf("matchRequirements cannot be empty")))
+	}
+	return
+}
+
+func validatePodMarkerMatchPreferences(preferences []appsv1alpha1.PodMarkerPreference, fldPath *field.Path) (allErrs field.ErrorList) {
+	if len(preferences) > LimitedPreferencesNumber {
+		allErrs = append(allErrs, field.Invalid(fldPath, preferences, fmt.Sprintf("over the limitation of the number of MatchPreferences, limitation: %d, yours: %d", LimitedPreferencesNumber, len(preferences))))
+	}
+	for i := range preferences {
+		preference := &preferences[i]
+		// 1. validate podSelector
+		if _, err := v1.LabelSelectorAsSelector(preference.PodSelector); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child(fmt.Sprintf("[%d].PodSelector", i)), *preference.PodSelector, fmt.Sprintf("invalid podSelector, err: %v", err)))
+		}
+		// 2. validate nodeSelector
+		if _, err := v1.LabelSelectorAsSelector(preference.NodeSelector); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child(fmt.Sprintf("[%d].NodeSelector", i)), *preference.PodSelector, fmt.Sprintf("invalid NodeSelector, err: %v", err)))
+		}
+	}
+	return
+}
+
+func validatePodMarkerMarkIdentities(markItems *appsv1alpha1.PodMarkerMarkItems, fldPath *field.Path) (allErrs field.ErrorList) {
+	// markItems cannot be empty
+	if len(markItems.Labels)+len(markItems.Annotations) == 0 {
+		return append(allErrs, field.Invalid(fldPath, *markItems, fmt.Sprintf("markItems cannot be empty")))
+	}
+	// 1. validate the labels users want to mark
+	allErrs = metav1validation.ValidateLabels(markItems.Labels, fldPath.Child("labels"))
```

Comment:
is possible the marked labels will make match selector invalid?

Reply:
sounds very reasonable